react_juce/core/TextInputView.cpp

```diff
@@ -0,0 +1,225 @@
+        {
+            // Invoke JavaScript's `change` event.
+            if ((*props).contains(TextInputView::onChangeProp) && (*props)[TextInputView::onChangeProp].isMethod())
+            {
+                std::array<juce::var, 1> args{{detail::makeChangeEventObject(getText())}};
+                juce::var::NativeFunctionArgs nfArgs(juce::var(), args.data(), static_cast<int>(args.size()));
+                std::invoke((*props)[TextInputView::onChangeProp].getNativeFunction(), nfArgs);
+            }
+            dirty = false;
+        }
+    }
+
+    //==============================================================================
+
+    TextInputView::TextInputView()
+        : textInput(&props)
+    {
+        addAndMakeVisible(textInput);
+        textInput.addListener(&textInput);
+        textInput.setPopupMenuEnabled(false);
+    }
+
+    void TextInputView::setProperty(const juce::Identifier &name, const juce::var &value)
+    {
+        View::setProperty(name, value);
+        if (name == valueProp)
+        {
+            if (!value.isString())
+                throw std::invalid_argument("Invalid prop value. Prop \'value\' must be a string.");
+            textInput.setControlled(true);
+            textInput.setControlledValue(value);
+        }
+        if (name == placeholderProp)
+        {
+            if (!value.isString())
+                throw std::invalid_argument("Invalid prop value. Prop \'placeholder\' must be a string.");
+            textInput.setPlaceholderText(value);
+        }
+        if (name == placeholderColorProp)
+        {
+            if (!value.isString())
+                throw std::invalid_argument("Invalid prop value. Prop \'placeholder-color\' must be a color string.");
+            juce::String hexPlaceholderColor = value;
+            juce::Colour placeholderColor = juce::Colour::fromString(hexPlaceholderColor);
+            textInput.setPlaceholderColour(placeholderColor);
+        }
+        if (name == maxlengthProp)
+        {
+            if (!value.isDouble())
+              throw std::invalid_argument("Invalid prop value. Prop \'maxlength\' must be a number.");
+            textInput.setMaxLength(value);
+        }
+        if (name == readonly)
+        {
+            textInput.setReadOnly(value);
+        }
+
+        textInput.applyFontToAllText(getFont());
+
+        juce::String hexColor = props.getWithDefault(colorProp, "ff000000");
+        juce::Colour colour = juce::Colour::fromString(hexColor);
+        textInput.applyColourToAllText(colour);
+
+        int just = props.getWithDefault(justificationProp, 1);
+        textInput.setJustification(just);
+
+        juce::String hexBackgroundColor = props.getWithDefault(backgroundColorProp, "00000000");
+        juce::Colour backgroundColour = juce::Colour::fromString(hexBackgroundColor);
+        textInput.setColour(juce::TextEditor::ColourIds::backgroundColourId, backgroundColour);
+
+        if (name == outlineColorProp)
```

Comment:
It feels like there might be a nice little refactor here to avoid some of this color setter duplication. 

I'm thinking maybe some sort of map of propID => setter which does the following code block a little more generically:

```
juce::String hexFocusedOutlineColor = value;
            juce::Colour focusedOutlineColor = juce::Colour::fromString(hexFocusedOutlineColor);
            textInput.setColour(juce::TextEditor::ColourIds::focusedOutlineColourId, focusedOutlineColor);
```

Reply:
I made a change to remove duplication (d7c2299).
Also I stopped providing default value to 'backgound-color' for simplicity.
Now I think if we really want default values, it is better to add default values to all props in `textEditorColourIdsByProp`.
What do you think @JoshMarler?

Reply:
Yeah. I'd even be tempted to make such things default prop values in React so users of the package can clearly see what the default color/control values will be. What do you think @nick-thompson ?

Reply:
Following up on default prop values in React components: yea, I agree that we should generally apply default properties at the React layer. For example

```js
function TextInput(props) {
    const defaultProps = {
        ...our defaults here...
    };
    return React.createElement('TextInput', Object.assign({}, defaultProps, props), props.children);
}
```

That way we can set defaults and change defaults re-using all the same code paths that we're used to.

I want to say too, though, that we should _still_ use `props.getWithDefault` when we're trying to read property values inside our `setProperty` overrides. The reason for that is that `setViewProperty` gets called many times during a React update batch, and so our `setProperty` overrides get called many times as well. For any given invocation of `setProperty`, we have no guarantees that _all_ of our properties have been set already. So even if we're relying on defaults at the React layer, we have to be prepared for the fact that we might not have all of our properties during any given invocation of a `setProperty` override, hence my suggestion that we should be using `props.getWithDefault` there generally.

This behavior can change slightly once we introduce the StyleSheet API, which I'll open an issue for now.

That all said, I think we're good to merge this PR!